src/dbnode/auth/inbound.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package auth
+
+import (
+	"fmt"
+
+	"github.com/uber/tchannel-go/thrift"
+)
+
+// Inbound encapsulates client credentials.
+type Inbound struct {
+	clientCredentials []InboundCredentials
+	authMode          Mode
+}
+
+// ValidateCredentials validates the inbound credential and return error accordingly.
+func (i *Inbound) ValidateCredentials(creds InboundCredentials) error {
+	if i == nil || i.authMode == AuthModeNoAuth {
+		return nil
+	}
+
+	if i.authMode == AuthModeShadow {
+		go func() {
+			credentialMatched := false
+			for _, p := range i.clientCredentials {
+				if i.MatchCredentials(p, creds) {
+					credentialMatched = true
+					break
+				}
+			}
+
+			if !credentialMatched {
+				// todo emit a metric
+			}
+		}()
+		return nil
+	}
+	if creds.Type == Unknown {
+		return fmt.Errorf("unknown credential type for dbnode inbound")
+	}
+
+	if creds.Type != ClientCredential {
+		return fmt.Errorf("incorrect credential type for dbnode inbound")
+	}
+
+	if err := creds.Validate(); err != nil {
+		return err
+	}
+	for _, p := range i.clientCredentials {
+		if i.MatchCredentials(p, creds) {
+			return nil
+		}
+	}
+	return fmt.Errorf("credential not matched for dbnode inbound")
+}
+
+// ValidateCredentialsFromThriftContext validates inbound credential from thrift context.
+func (i *Inbound) ValidateCredentialsFromThriftContext(tctx thrift.Context, credtype CredentialType) error {
+	ctxHeaders := tctx.Headers()
+	userName, ok := ctxHeaders[AuthUsername]
```

Comment:
I think we discussed it offline, but why don't we encrypt the username/password before sending it as a header? We provide user name/password to the node at start up, we might as well pass a private part of asymetric key. User encrypt the username/password using public part before sending the request and we decrypt it here.
I don't think it will add a lot of overhead but without it sending username/password in plain text as a header doesn't provide a lot of security